packages/medusa/src/services/payment-collection.ts

```diff
@@ -0,0 +1,162 @@
+
+import { CreatePaymentCollectionInput } from "../types/payment-collection"
+
+type InjectedDependencies = {
+  manager: EntityManager
+  paymentCollectionRepository: typeof PaymentCollectionRepository
+  eventBusService: EventBusService
+}
+
+export default class PaymentCollectionService extends TransactionBaseService {
+  static readonly Events = {
+    CREATED: "payment-collection.created",
+    UPDATED: "payment-collection.updated",
+    DELETED: "payment-collection.deleted",
+  }
+
+  protected readonly manager_: EntityManager
+  protected transactionManager_: EntityManager | undefined
+  protected readonly eventBusService_: EventBusService
+  // eslint-disable-next-line max-len
+  protected readonly paymentCollectionRepository_: typeof PaymentCollectionRepository
+
+  constructor({
+    manager,
+    paymentCollectionRepository,
+    eventBusService,
+  }: InjectedDependencies) {
+    // eslint-disable-next-line prefer-rest-params
+    super(arguments[0])
+
+    this.manager_ = manager
+    this.paymentCollectionRepository_ = paymentCollectionRepository
+    this.eventBusService_ = eventBusService
+  }
+
+  async retrieve(
+    paymentCollectionId: string,
+    config: FindConfig<PaymentCollection> = {}
+  ): Promise<PaymentCollection> {
+    const manager = this.transactionManager_ ?? this.manager_
+    const paymentCollectionRepository = manager.getCustomRepository(
+      this.paymentCollectionRepository_
+    )
+
+    const query = buildQuery({ id: paymentCollectionId }, config)
+    const paymentCollection = await paymentCollectionRepository.findOne(query)
+
+    if (!paymentCollection) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Payment collection with id ${paymentCollectionId} was not found`
+      )
+    }
+
+    return paymentCollection
+  }
+
+  async create(
+    data: CreatePaymentCollectionInput,
+    context: { loggedInUserId: string }
+  ): Promise<PaymentCollection> {
+    return await this.atomicPhase_(async (transactionManager) => {
+      const paymentCollectionRepository =
+        transactionManager.getCustomRepository(
+          this.paymentCollectionRepository_
+        )
+
+      const paymentCollectionToCreate = paymentCollectionRepository.create({
+        region_id: data.region_id,
+        type: data.type,
+        status: PaymentCollectionStatus.NOT_PAID,
+        currency_code: data.currency_code,
+        amount: data.amount,
+        metadata: data.metadata,
+        created_by: context.loggedInUserId,
+        description: data.description,
+      })
+
+      const paymentCollection = await paymentCollectionRepository.save(
+        paymentCollectionToCreate
+      )
+
+      await this.eventBusService_
+        .withTransaction(transactionManager)
+        .emit(PaymentCollectionService.Events.CREATED, paymentCollection)
+
+      return paymentCollection
+    })
+  }
+
+  async update(
+    paymentCollectionId: string,
+    data: DeepPartial<PaymentCollection>
```

Comment:
suggestion: Add a `UpdatePaymentCollectionInput` enforcing a type check for fields that are not allowed to be updated e.g. `created_by`

Reply:
This `update` will be used internally in multiple other methods, so the plan was to add these checks at the other methods as each of them will be restricted to a few fields. In theory that method wouldn't be called directly from an endpoint for example.